File: src/providers/droid.ts
import { readFileSync, existsSync, readdirSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import type { SkillCall } from "../models.js";
import type { Provider } from "./base.js";

const FACTORY_HOME = join(homedir(), ".factory");
const SESSIONS_DIR = join(FACTORY_HOME, "sessions");

const SKILL_ACTIVE_RE = /Skill "([^"]+)" is now active/;

export class DroidProvider implements Provider {
  readonly name = "Droid CLI";

  available(): boolean {
    return existsSync(SESSIONS_DIR);
  }

  collect(): SkillCall[] {
    if (!this.available()) return [];

    const calls: SkillCall[] = [];
    this.walkSessions(SESSIONS_DIR, calls);
    return calls;
  }

  private walkSessions(dir: string, calls: SkillCall[]) {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const fullPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        this.walkSessions(fullPath, calls);
      } else if (entry.name.endsWith(".jsonl")) {
        this.parseSession(fullPath, dir, calls);
      }
    }
  }

  private parseSession(path: string, parentDir: string, calls: SkillCall[]) {
    let content: string;
    try {
      content = readFileSync(path, "utf-8");
    } catch {
      return;
    }

    // Derive project from directory structure: sessions/{project-slug}/session.jsonl
    const relDir = parentDir.replace(SESSIONS_DIR, "").replace(/^\//, "");
    const project = relDir ? "/" + relDir.replace(/-/g, "/") : "";

    const lines = content.split("\n");
    let sessionId = "";

    for (const line of lines) {
      if (!line.trim()) continue;
      let entry: any;
      try {
        entry = JSON.parse(line);
      } catch {
        continue;
      }

      if (entry.type === "session_start") {
        sessionId = entry.id ?? "";
      }

      if (entry.type !== "message") continue;
      const msg = entry.message;
      if (!msg) continue;

      const msgContent = msg.content;
      if (!Array.isArray(msgContent)) continue;

      for (const part of msgContent) {
        if (part.type !== "tool_result") continue;
        const text: string = typeof part.content === "string"
          ? part.content
          : "";
        const match = text.match(SKILL_ACTIVE_RE);
        if (!match) continue;

        const skill = match[1]!;
        const ts = entry.timestamp
          ? new Date(entry.timestamp)
          : new Date(0);

        calls.push({
          skill,
          timestamp: ts,
          project,
          sessionId,
          source: this.name,
        });
      }
    }
  }
}

File: index/src/providers/droid.rs
use std::path::Path;

use regex::Regex;
use serde_json::Value;
use walkdir::WalkDir;

use crate::model::{ProviderResult, SkillCall};

const SOURCE: &str = "Droid CLI";

pub fn collect(home: &str) -> ProviderResult {
    let sessions_dir = format!("{home}/.factory/sessions");
    let available = Path::new(&sessions_dir).is_dir();

    if !available {
        return ProviderResult {
            name: SOURCE.into(),
            available: false,
            calls: vec![],
        };
    }

    let active_re = Regex::new(r#"Skill "([^"]+)" is now active"#).unwrap();
    let mut calls = Vec::new();

    for entry in WalkDir::new(&sessions_dir)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
            continue;
        }

        parse_session(path, &active_re, &mut calls);
    }

    ProviderResult {
        name: SOURCE.into(),
        available,
        calls,
    }
}

fn parse_session(path: &Path, active_re: &Regex, calls: &mut Vec<SkillCall>) {
    let content = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return,
    };

    let mut session_id = String::new();
    let mut project = String::new();

    for line in content.lines() {
        if line.trim().is_empty() {
            continue;
        }

        let entry: Value = match serde_json::from_str(line) {
            Ok(v) => v,
            Err(_) => continue,
        };

        if entry["type"].as_str() == Some("session_start") {
            session_id = entry["id"].as_str().unwrap_or("").to_string();
            // Extract project path from session metadata
            if let Some(cwd) = entry["cwd"].as_str() {
                project = cwd.to_string();
            }
            continue;
        }

        if entry["type"].as_str() != Some("message") {
            continue;
        }

        let content_arr = match entry["message"]["content"].as_array() {
            Some(a) => a,
            None => continue,
        };

        for part in content_arr {
            if part["type"].as_str() != Some("tool_result") {
                continue;
            }

            let text = part["content"].as_str().unwrap_or("");
            let caps = match active_re.captures(text) {
                Some(c) => c,
                None => continue,
            };
            let skill = &caps[1];

            let ts = entry["timestamp"]
                .as_str()
                .and_then(super::claude_code::parse_iso_ms)
                .or_else(|| entry["timestamp"].as_i64())
                .unwrap_or(0);

            calls.push(SkillCall {
                skill: skill.to_string(),
                timestamp_ms: ts,
                project: project.clone(),
                session_id: session_id.clone(),
                source: SOURCE.into(),
            });
        }
    }
}
